File: backend/app/services/profiles.py
"""First-class generation profiles. Public names stay professional."""

from __future__ import annotations

from typing import Any

# Interactive Dialogue defaults are anchored on the Qwen3.5 non-thinking card,
# then adjusted from local A/B (see benchmarks/dialogue_reliability/).
# repetition_penalty uses HuggingFace/Qwen scale: 1.0 means off.
INTERACTIVE_DIALOGUE = {
    "profile": "interactive_dialogue",
    "enable_thinking": False,
    "reasoning_effort": "medium",
    "thinking_continuation": False,
    "max_tokens": 1536,
    "temperature": 0.7,
    "top_p": 0.8,
    "top_k": 20,
    "min_p": 0.0,
    "presence_penalty": 0.5,
    "presence_context_size": 256,
    "frequency_penalty": 0.0,
    "frequency_context_size": 256,
    "repetition_penalty": 1.0,
    "repetition_context_size": 128,
    "prompt_soft_target": 8192,
    "prompt_budget": 10240,
}

BALANCED = {
    **INTERACTIVE_DIALOGUE,
    "profile": "balanced",
    "max_tokens": 4096,
    "presence_penalty": 0.0,
    "temperature": 0.7,
    "top_p": 0.9,
    "prompt_soft_target": 12288,
    "prompt_budget": 16384,
}

REASONING = {
    "profile": "reasoning",
    "enable_thinking": True,
    "reasoning_effort": "medium",
    "thinking_continuation": False,
    "max_tokens": 8192,
    "temperature": 0.6,
    "top_p": 0.95,
    "top_k": 20,
    "min_p": 0.0,
    "presence_penalty": 0.0,
    "presence_context_size": 20,
    "frequency_penalty": 0.0,
    "frequency_context_size": 20,
    "repetition_penalty": 1.0,
    "repetition_context_size": 20,
    "prompt_soft_target": 16384,
    "prompt_budget": 32768,
}

# Long-form narrative: ordinary chat stays on interactive_dialogue.
# Per-segment max_tokens is a budget, not the whole-work limit.
LONG_FORM = {
    **BALANCED,
    "profile": "long_form",
    "max_tokens": 2048,
    "segment_max_tokens": 2048,
    "target_visible_chars": 20000,
    "segment_chars": 2500,
    "mode": "narrative",
    "enable_thinking": False,
    "presence_penalty": 0.3,
    "prompt_soft_target": 12288,
    "prompt_budget": 16384,
}

PROFILES: dict[str, dict[str, Any]] = {
    "interactive_dialogue": INTERACTIVE_DIALOGUE,
    "conversational": INTERACTIVE_DIALOGUE,
    "balanced": BALANCED,
    "reasoning": REASONING,
    "long_form": LONG_FORM,
    "narrative": LONG_FORM,
}

DEFAULT_PROFILE = "interactive_dialogue"


def normalize_profile(name: str | None) -> str:
    if not name:
        return DEFAULT_PROFILE
    key = name.strip().lower().replace("-", "_").replace(" ", "_")
    aliases = {
        "dialogue": "interactive_dialogue",
        "chat": "interactive_dialogue",
        "think": "reasoning",
        "thinking": "reasoning",
        "longform": "long_form",
        "long_output": "long_form",
        "story": "long_form",
    }
    key = aliases.get(key, key)
    if key not in PROFILES:
        return DEFAULT_PROFILE
    if key == "conversational":
        return "interactive_dialogue"
    if key == "narrative":
        return "long_form"
    return key


def resolve_profile(name: str | None) -> dict[str, Any]:
    return dict(PROFILES[normalize_profile(name)])

File: web/src/lib/profiles.ts
import type { GenerationParams, GenerationProfile } from "../types/chat";

export const PROFILE_PRESETS: Record<GenerationProfile, GenerationParams> = {
  interactive_dialogue: {
    profile: "interactive_dialogue",
    temperature: 0.7,
    topP: 0.8,
    topK: 20,
    minP: 0,
    presencePenalty: 0.5,
    presenceContextSize: 256,
    frequencyPenalty: 0,
    frequencyContextSize: 256,
    repetitionPenalty: 1.0,
    repetitionContextSize: 128,
    maxTokens: 1536,
    enableThinking: false,
    reasoningEffort: "medium",
  },
  balanced: {
    profile: "balanced",
    temperature: 0.7,
    topP: 0.9,
    topK: 20,
    minP: 0,
    presencePenalty: 0,
    presenceContextSize: 256,
    frequencyPenalty: 0,
    frequencyContextSize: 256,
    repetitionPenalty: 1.0,
    repetitionContextSize: 128,
    maxTokens: 4096,
    enableThinking: false,
    reasoningEffort: "medium",
  },
  reasoning: {
    profile: "reasoning",
    temperature: 0.6,
    topP: 0.95,
    topK: 20,
    minP: 0,
    presencePenalty: 0,
    presenceContextSize: 20,
    frequencyPenalty: 0,
    frequencyContextSize: 20,
    repetitionPenalty: 1.0,
    repetitionContextSize: 20,
    maxTokens: 8192,
    enableThinking: true,
    reasoningEffort: "medium",
  },
  long_form: {
    profile: "long_form",
    temperature: 0.7,
    topP: 0.9,
    topK: 20,
    minP: 0,
    presencePenalty: 0.3,
    presenceContextSize: 256,
    frequencyPenalty: 0,
    frequencyContextSize: 256,
    repetitionPenalty: 1.0,
    repetitionContextSize: 128,
    maxTokens: 2048,
    enableThinking: false,
    reasoningEffort: "medium",
  },
};

export const PROFILE_LABELS: Record<GenerationProfile, string> = {
  interactive_dialogue: "Interactive Dialogue",
  balanced: "Balanced",
  reasoning: "Reasoning",
  long_form: "Long Form (20K+)",
};

export function isIncompleteTerminal(finish?: string | null, terminal?: string | null): boolean {
  const state = terminal || finish || "";
  return [
    "interrupted_transport",
    "interrupted_user",
    "upstream_protocol_error",
    "timeout",
    "generation_error",
    "repetition_guard",
    "unknown_terminal",
    "completed_with_transport_error",
    "error",
    "abort",
  ].includes(state);
}

export function terminalCopy(finish?: string | null, terminal?: string | null): string | null {
  const state = terminal || finish || "";
  if (state === "length" || state === "completed_length") return "已达到输出上限";
  if (state === "repetition_guard") return "检测到明显重复，已停止生成";
  if (state === "timeout") return "生成超时";
  if (state === "interrupted_user" || state === "abort") return "已中断";
  if (state === "completed_with_transport_error") return "传输有损坏，这段回复可能缺字";
  if (
    state === "interrupted_transport" ||
    state === "upstream_protocol_error" ||
    state === "unknown_terminal" ||
    state === "generation_error"
  ) {
    return "生成未正常完成";
  }
  return null;
}
